mono/utils/mono-state.c

```diff
@@ -26,263 +31,280 @@ extern GCStats mono_gc_stats;
-	memset (output_dump_str, 0, len * sizeof (gchar));
+	// Each individual key/value append should be roughly less than this many characters
+	const int margin = 35;
 
-	static_gstr.len = 0;
-	static_gstr.allocated_len = len;
-	static_gstr.str = output_dump_str;
+	// Not using static, exit
+	if (writer->allocated_len == 0)
+		return;
 
-	writer.indent = 0;
-	writer.text = &static_gstr;
+	if (writer->allocated_len - writer->len < margin)
+		g_error ("Ran out of memory to create crash dump json blob. Current state:\n%s\n", writer->output_str);
 }
 
-static void mono_json_writer_init_with_static (void) 
+static void 
+mono_state_writer_indent (MonoStateWriter *writer)
 {
-	return mono_json_writer_init_memory (output_dump_str, MONO_MAX_SUMMARY_LEN);
+	for (int i = 0; i < writer->indent; ++i)
+		FORMAT_APPEND_STATIC(writer, " ");
 }
 
-static void assert_has_space (void)
+static void 
+mono_state_writer_object_key (MonoStateWriter *writer, const char *key) 
 {
-	// Each individual key/value append should be roughly less than this many characters
-	const int margin = 35;
-
-	// Not using static, exit
-	if (static_gstr.allocated_len == 0)
-		return;
-
-	if (static_gstr.allocated_len - static_gstr.len < margin)
-		g_error ("Ran out of memory to create crash dump json blob. Current state:\n%s\n", static_gstr.str);
+	mono_state_writer_indent (writer);
+	FORMAT_APPEND_STATIC(writer, "\"%s\" : ", key);
 }
 
 static void
-mono_native_state_add_ctx (JsonWriter *writer, MonoContext *ctx)
+mono_native_state_add_ctx (MonoStateWriter *writer, MonoContext *ctx)
 {
 	// Context
-	mono_json_writer_indent (writer);
-	mono_json_writer_object_key(writer, "ctx");
-	mono_json_writer_object_begin(writer);
-
-	assert_has_space ();
-	mono_json_writer_indent (writer);
-	mono_json_writer_object_key(writer, "IP");
-	mono_json_writer_printf (writer, "\"%p\",\n", (gpointer) MONO_CONTEXT_GET_IP (ctx));
-
-	assert_has_space ();
-	mono_json_writer_indent (writer);
-	mono_json_writer_object_key(writer, "SP");
-	mono_json_writer_printf (writer, "\"%p\",\n", (gpointer) MONO_CONTEXT_GET_SP (ctx));
-
-	assert_has_space ();
-	mono_json_writer_indent (writer);
-	mono_json_writer_object_key(writer, "BP");
-	mono_json_writer_printf (writer, "\"%p\"\n", (gpointer) MONO_CONTEXT_GET_BP (ctx));
-
-	mono_json_writer_indent_pop (writer);
-	mono_json_writer_indent (writer);
-	mono_json_writer_object_end (writer);
-	mono_json_writer_printf (writer, ",\n");
+	mono_state_writer_indent (writer);
+	mono_state_writer_object_key (writer, "ctx");
+	FORMAT_APPEND_STATIC(writer, "{\n");
+	writer->indent++;
+
+	assert_has_space (writer);
+	mono_state_writer_indent (writer);
+	mono_state_writer_object_key (writer, "IP");
+	FORMAT_APPEND_STATIC(writer, "\"%p\",\n", (gpointer) MONO_CONTEXT_GET_IP (ctx));
+
+	assert_has_space (writer);
+	mono_state_writer_indent (writer);
+	mono_state_writer_object_key (writer, "SP");
+	FORMAT_APPEND_STATIC(writer, "\"%p\",\n", (gpointer) MONO_CONTEXT_GET_SP (ctx));
```

Comment:
Consider a different approach -- sprintf is a slow, type unsafe, overly general function. Very popular, but still slow, unsafe, overly  general. The slowness doesn't matter in crashing, but the rest maybe.
```
json_append_char  (writer, '"');
json_append_pointer  (writer, (gpointer) MONO_CONTEXT_GET_SP (ctx));
json_append_char  (writer, '"');
```

Reply:
@jaykrell aren't you still going to end up needing to use format strings for all the "%x" "%p" "0x%llx" values? Or are you suggesting manual formatting of all of these strings?

Reply:
That is one approach, and it does add a sort of type safety, but no, you can write a bunch of small functions that are simpler and faster.

```
#include <stdio.h>
#include <stddef.h>
typedef int gboolean;
#define TRUE 1
#define FALSE 0
typedef unsigned long long uint64;
typedef long long int64;

void
mono_strrev (char* buf, size_t len)
{
 if (len < 2) return;
 size_t j = len - 1;
 for (size_t i = 0; i < j; ++i, --j)
 {
   char c = buf [i];
   buf [i] = buf [j];
   buf [j] = c;
 }
}

int
mono_sprintf_int (char* buf, uint64 u, unsigned base, gboolean upper, unsigned width)
{
int len = 0;
do
{
 buf [len++] = (upper ? "0123456789ABCDEF" : "0123456789abcdef") [u % base];
 width -= (width > 0);
 u /= base;
} while (u);
while (width--)
  buf [len++] = '0';
 mono_strrev (buf, len);
 return len;
}

int
mono_sprintf_u (char* buf, uint64 u)
{
 return mono_sprintf_int (buf, u, 10, FALSE, 0);
}

int
mono_sprintf_d(char* buf, int64 i)
{
 uint64 u;
 int neg = (i < 0);
 if (neg) {
    *buf++ = '-';
    u = 1 + (uint64)-(i + 1); // avoid negating the most negative number
 }
 else u = (uint64)i;
 return neg + mono_sprintf_int (buf, u, 10, FALSE, 0);
}

int
mono_sprintf_x (char* buf, uint64 u)
{
 return mono_sprintf_int (buf, u, 16, FALSE, 0);
}

int
mono_sprintf_X (char* buf, uint64 u)
{
 return mono_sprintf_int (buf, u, 16, TRUE, 0);
}

int
mono_sprintf_8x (char* buf, uint64 u)
{
 return mono_sprintf_int (buf, u, 16, FALSE, 8);
}


int main()
{
char  buf[99];

memset (buf,  'X', sizeof (buf));
buf [mono_sprintf_int (buf, 0x123, 16, 0, 0)] = 0;
printf("0x%s\n", buf);

memset (buf,  'X', sizeof (buf));
buf [mono_sprintf_int (buf, 123, 10, 0, 0)] = 0;
printf("%s\n", buf);

memset (buf,  'X', sizeof (buf));
buf [mono_sprintf_int (buf, 0x123a, 16, 1, 16)] = 0;
printf("%s\n", buf);

memset (buf,  'X', sizeof (buf));
buf [mono_sprintf_int (buf, 0x1234ABCD1234ABCD, 16, 1, 16)] = 0;
printf("%s\n", buf);

memset (buf,  'X', sizeof (buf));
buf [mono_sprintf_d (buf, -123)] = 0;
printf("%s\n", buf);

memset (buf,  'X', sizeof (buf));
buf [mono_sprintf_u (buf, 123)] = 0;
printf("%s\n", buf);
}

```

Reply:
sprintf also has reentrancy problems, at least on some  platforms.
On Visual  C++ it deals with locales and locks, and %f does a heap allocation.
Perhaps the non-Windows versions are more reentrant.